dap_prinz_green_jobs/pipeline/green_measures/occupations/soc_map.py

```diff
@@ -0,0 +1,378 @@
+                )
+            else:
+                self.all_soc_embeddings = load_s3_data(BUCKET_NAME, embeddings_path)
+                self.soc_job_titles = load_s3_data(BUCKET_NAME, job_titles_path)
+        except:
+            logger.info(f"SOC job title embeddings not found")
+
+            # Embed the SOC job titles
+            self.soc_job_titles = list(self.job_title_2_soc6_4.keys())
+
+            self.all_soc_embeddings = self.embed_texts(self.soc_job_titles)
+
+            if save_embeds:
+                logger.info(f"Saving SOC job title embeddings")
+                save_to_s3(BUCKET_NAME, self.all_soc_embeddings, embeddings_path)
+                save_to_s3(BUCKET_NAME, self.soc_job_titles, job_titles_path)
+
+    def find_most_similar_matches(
+        self,
+        job_titles: Union[str, List[str]],
+        job_title_embeddings: np.array(object),
+    ) -> dict:
+        """
+        Using the job title embeddings and the SOC job title embeddings,
+        find the top n SOC job titles which are most similar to each input job title.
+        """
+
+        logger.info(f"Finding most similar job titles for {len(job_titles)} job titles")
+
+        similarities = cosine_similarity(job_title_embeddings, self.all_soc_embeddings)
+
+        # Top matches for each data point
+        job_top_soc_matches = []
+        for job_title_ix, job_title in tqdm(enumerate(job_titles)):
+            top_soc_matches = []
+            for soc_ix in np.flip(np.argsort(similarities[job_title_ix]))[
+                0 : self.match_top_n
+            ]:
+                soc_text = self.soc_job_titles[soc_ix]
+                top_soc_matches.append(
+                    [
+                        soc_text,
+                        self.job_title_2_soc6_4[soc_text][0],  # 6 digit
+                        self.job_title_2_soc6_4[soc_text][1],  # 4 digit
+                        similarities[job_title_ix][soc_ix],
+                    ]
+                )
+            job_top_soc_matches.append(
+                {
+                    "job_title": job_title,
+                    "top_soc_matches": top_soc_matches,
+                }
+            )
+
+        return job_top_soc_matches
+
+    def find_most_likely_soc(
+        self,
+        match_row: dict,
+    ) -> tuple:
+        """
+        For a single job title and the details of the most similar SOC matches, find a single most likely SOC
+        1. If the top match has a really high similarity score (>sim_threshold) at the 6-digit level then use this.
+                This will return (soc, job_title)
+        2. Get the 4-digit SOCs of the good (>top_n_sim_threshold) matches in the top n most similar.
+        3. If there are a few of these (>=minimum_n) and over a certain proportion (>minimum_prop) of these are the same at the 4 digit level - use this as the SOC.
+                This will return (soc, the job titles given for this same soc)
+        """
```

Comment:
makes a ton of sense as an approach and looks like the outputs make a lot of sense - it would be interesting to analyse the job titles we can't match to in our sample to see if there are any systematic patterns in non-matching (i.e. better text cleaning or the job title itself is bad)